flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -485,76 +481,152 @@ public boolean isShutdown() {
+			externalSavepointLocation,
+			isPeriodic,
+			advanceToEndOfTime,
+			onCompletionPromise);
+		return onCompletionPromise;
+	}
 
-		// next, check if all tasks that need to acknowledge the checkpoint are running.
-		// if not, abort the checkpoint
-		Map<ExecutionAttemptID, ExecutionVertex> ackTasks = new HashMap<>(tasksToWaitFor.length);
+	private void startTriggeringCheckpoint(
+		long timestamp,
+		CheckpointProperties props,
+		@Nullable String externalSavepointLocation,
+		boolean isPeriodic,
+		boolean advanceToEndOfTime,
+		CompletableFuture<CompletedCheckpoint> onCompletionPromise) {
 
-		for (ExecutionVertex ev : tasksToWaitFor) {
-			Execution ee = ev.getCurrentExecutionAttempt();
-			if (ee != null) {
-				ackTasks.put(ee.getAttemptId(), ev);
-			} else {
-				LOG.info("Checkpoint acknowledging task {} of job {} is not being executed at the moment. Aborting checkpoint.",
-						ev.getTaskNameWithSubtaskIndex(),
-						job);
-				throw new CheckpointException(CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
+		try {
+			// make some eager pre-checks
+			synchronized (lock) {
+				preCheckBeforeTriggeringCheckpoint(isPeriodic, props.forceCheckpoint());
 			}
-		}
 
-		// we will actually trigger this checkpoint!
+			final Execution[] executions = getTriggerExecutions();
+			final Map<ExecutionAttemptID, ExecutionVertex> ackTasks = getAckTasks();
+
+			// we will actually trigger this checkpoint!
+			Preconditions.checkState(!isTriggering);
+			isTriggering = true;
+
+			final CompletableFuture<PendingCheckpoint> pendingCheckpointCompletableFuture =
+				initializeCheckpoint(props, externalSavepointLocation)
+					.thenApplyAsync(
+						(checkpointIdAndStorageLocation) -> createPendingCheckpoint(
+							timestamp,
+							props,
+							ackTasks,
+							isPeriodic,
+							checkpointIdAndStorageLocation.checkpointId,
+							checkpointIdAndStorageLocation.checkpointStorageLocation,
+							onCompletionPromise),
+						timer);
+
+			pendingCheckpointCompletableFuture
+				.thenCompose(this::snapshotMasterState)
+				.whenCompleteAsync(
+					(ignored, throwable) -> {
+						final PendingCheckpoint checkpoint =
+							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
+
+						if (throwable == null && checkpoint != null && !checkpoint.isDiscarded()) {
+							// no exception, no discarding, everything is OK
+							snapshotTaskState(
+								timestamp,
+								checkpoint.getCheckpointId(),
+								checkpoint.getCheckpointStorageLocation(),
+								props,
+								executions,
+								advanceToEndOfTime);
+							onTriggerSuccess();
+						} else {
+								// the initialization might not be finished yet
+								if (checkpoint == null) {
+									onTriggerFailure(onCompletionPromise, throwable);
+								} else {
+									onTriggerFailure(checkpoint, throwable);
+								}
+						}
+					},
+					timer);
+		} catch (Throwable throwable) {
+			onTriggerFailure(onCompletionPromise, throwable);
```

Comment:
Hmmm, are all of the invocations of `#startTriggeringCheckpoint(...)` coming from the `timer` thread? I think so, but If not, there would be a potential for race conditions between the calls to `onTriggerFailure()` (here) coming from that other thread and the timer thread.

Reply:
Yes, `startTriggeringCheckpoint` must be executed in timer thread. We can guarantee that now.